File: examples/data-science/001-dataframe-ops/implementations/julia/dataframe_ops.jl
#!/usr/bin/env julia
"""
DataFrame Operations - Julia Implementation
High-performance scientific computing equivalent to Ruchy implementation
Focus: Type-safe numerical operations with performance optimization
"""

using Statistics
using BenchmarkTools

function create_test_data()::Vector{Int}
    """Create test data matching Ruchy implementation"""
    return [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
end

function calculate_sum(data::Vector{Int})::Int
    """Calculate sum of values"""
    return sum(data)
end

function calculate_mean(data::Vector{Int})::Int
    """Calculate mean of values (integer division to match Ruchy)"""
    if length(data) == 0
        return 0
    end
    return sum(data) ÷ length(data)
end

function filter_above_threshold(data::Vector{Int}, threshold::Int)::Vector{Int}
    """Filter values above threshold"""
    return filter(x -> x > threshold, data)
end

function sort_values(data::Vector{Int})::Vector{Int}
    """Sort values in ascending order"""
    return sort(data)
end

function find_maximum(data::Vector{Int})::Int
    """Find maximum value"""
    return length(data) > 0 ? maximum(data) : 0
end

function find_minimum(data::Vector{Int})::Int
    """Find minimum value"""
    return length(data) > 0 ? minimum(data) : 0
end

function count_consecutive_groups(data::Vector{Int})::Vector{Int}
    """Group consecutive values and count occurrences"""
    if length(data) == 0
        return Int[]
    end
    
    result = Int[]
    current_count = 1
    current_value = data[1]
    
    for i in 2:length(data)
        if data[i] == current_value
            current_count += 1
        else
            push!(result, current_count)
            current_value = data[i]
            current_count = 1
        end
    end
    
    push!(result, current_count)
    return result
end

function test_data_operations()
    """Test all basic data operations"""
    println("Basic Data Operations Tests - Julia")
    println("===================================")
    
    test_data = create_test_data()
    
    # Test 1: Sum calculation
    sum_result = calculate_sum(test_data)
    if sum_result == 55
        println("✓ Sum calculation: Pass")
    else
        println("✗ Sum calculation: Fail")
    end
    
    # Test 2: Mean calculation
    mean_result = calculate_mean(test_data)
    if mean_result == 5
        println("✓ Mean calculation: Pass")
    else
        println("✗ Mean calculation: Fail")
    end
    
    # Test 3: Filtering
    filtered = filter_above_threshold(test_data, 5)
    if length(filtered) == 5
        println("✓ Filtering operations: Pass")
    else
        println("✗ Filtering operations: Fail")
    end
    
    # Test 4: Sorting
    test_unsorted = [5, 2, 8, 1, 9]
    sorted_data = sort_values(test_unsorted)
    if sorted_data[1] == 1 && sorted_data[5] == 9
        println("✓ Sorting operations: Pass")
    else
        println("✗ Sorting operations: Fail")
    end
    
    # Test 5: Maximum/Minimum
    max_val = find_maximum(test_data)
    min_val = find_minimum(test_data)
    if max_val == 10 && min_val == 1
        println("✓ Min/Max operations: Pass")
    else
        println("✗ Min/Max operations: Fail")
    end
    
    # Test 6: Group counting
    grouped_data = [1, 1, 2, 2, 2, 3]
    counts = count_consecutive_groups(grouped_data)
    if length(counts) == 3 && counts[1] == 2 && counts[2] == 3 && counts[3] == 1
        println("✓ Group counting: Pass")
    else
        println("✗ Group counting: Fail")
    end
    
    println("")
    println("Basic data operations validation complete")
end

function benchmark_operations(iterations::Int = 10000)
    """Benchmark operations for performance comparison"""
    println("Benchmarking with $iterations iterations...")
    
    # Generate larger test data
    large_data = collect(1:1000)
    
    # Benchmark sum calculation
    sum_time = @elapsed for _ in 1:iterations
        calculate_sum(large_data)
    end
    
    # Benchmark sorting
    sort_time = @elapsed for _ in 1:iterations
        sort_values(copy(large_data))
    end
    
    # Benchmark filtering
    filter_time = @elapsed for _ in 1:iterations
        filter_above_threshold(large_data, 500)
    end
    
    println("Sum calculation: $(sum_time)s")
    println("Sorting: $(sort_time)s")
    println("Filtering: $(filter_time)s")
    
    return Dict(
        "sum_time" => sum_time,
        "sort_time" => sort_time,
        "filter_time" => filter_time
    )
end

function analyze_complexity()
    """Analyze complexity of operations"""
    println("Data Operations Complexity Analysis - Julia")
    println("==========================================")
    
    println("Operation Complexities:")
    println("  Sum/Mean calculation: O(n)")
    println("  Filtering: O(n)")
    println("  Sorting (quicksort): O(n log n)")
    println("  Min/Max finding: O(n)")
    println("  Group counting: O(n)")
    println("")
    
    println("Memory Complexity:")
    println("  Data storage: O(n)")
    println("  Operation results: O(k) where k ≤ n")
    println("")
    
    println("Julia Advantages:")
    println("  ✓ Just-in-time compilation to native code")
    println("  ✓ Near-C performance with high-level syntax")
    println("  ✓ Built for scientific computing")
    println("  ✓ Multiple dispatch system")
    println("  ✓ Excellent numerical stability")
    println("")
    
    println("Ruchy Advantages:")
    println("  ✓ Compile-time formal verification")
    println("  ✓ Static analysis and provability checking")
    println("  ✓ Memory safety without garbage collection")
    println("  ✓ Deterministic performance characteristics")
    println("  ✓ Zero-cost abstractions")
end

function main()
    """Main demonstration function"""
    println("Data Operations Foundation - Julia")
    println("=================================")
    println("High-performance scientific computing comparison with Ruchy")
    println("")
    
    # Run comprehensive test suite
    test_data_operations()
    println("")
    
    # Run benchmarks
    benchmark_results = benchmark_operations()
    println("")
    
    # Analyze complexity
    analyze_complexity()
    println("")
    
    println("✅ Julia implementation complete")
    println("📊 Ready for performance comparison with Ruchy")
    println("🎯 SPRINT 23: Cross-language validation in progress")
end

# Run main function when script is executed
if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: examples/data-science/001-dataframe-ops/implementations/r/dataframe_ops.R
#!/usr/bin/env Rscript
# DataFrame Operations - R Implementation
# Statistical computing equivalent to Ruchy implementation
# Focus: Statistical operations with R's built-in data analysis capabilities

library(dplyr)
library(microbenchmark)

create_test_data <- function() {
    # Create test data matching Ruchy implementation
    return(c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10))
}

calculate_sum <- function(data) {
    # Calculate sum of values
    return(sum(data))
}

calculate_mean <- function(data) {
    # Calculate mean of values (integer division to match Ruchy)
    if (length(data) == 0) {
        return(0)
    }
    return(as.integer(sum(data) / length(data)))
}

filter_above_threshold <- function(data, threshold) {
    # Filter values above threshold
    return(data[data > threshold])
}

sort_values <- function(data) {
    # Sort values in ascending order
    return(sort(data))
}

find_maximum <- function(data) {
    # Find maximum value
    if (length(data) == 0) {
        return(0)
    }
    return(max(data))
}

find_minimum <- function(data) {
    # Find minimum value
    if (length(data) == 0) {
        return(0)
    }
    return(min(data))
}

count_consecutive_groups <- function(data) {
    # Group consecutive values and count occurrences
    if (length(data) == 0) {
        return(c())
    }
    
    result <- c()
    current_count <- 1
    current_value <- data[1]
    
    if (length(data) > 1) {
        for (i in 2:length(data)) {
            if (data[i] == current_value) {
                current_count <- current_count + 1
            } else {
                result <- c(result, current_count)
                current_value <- data[i]
                current_count <- 1
            }
        }
    }
    
    result <- c(result, current_count)
    return(result)
}

test_data_operations <- function() {
    # Test all basic data operations
    cat("Basic Data Operations Tests - R\n")
    cat("===============================\n")
    
    test_data <- create_test_data()
    
    # Test 1: Sum calculation
    sum_result <- calculate_sum(test_data)
    if (sum_result == 55) {
        cat("✓ Sum calculation: Pass\n")
    } else {
        cat("✗ Sum calculation: Fail\n")
    }
    
    # Test 2: Mean calculation
    mean_result <- calculate_mean(test_data)
    if (mean_result == 5) {
        cat("✓ Mean calculation: Pass\n")
    } else {
        cat("✗ Mean calculation: Fail\n")
    }
    
    # Test 3: Filtering
    filtered <- filter_above_threshold(test_data, 5)
    if (length(filtered) == 5) {
        cat("✓ Filtering operations: Pass\n")
    } else {
        cat("✗ Filtering operations: Fail\n")
    }
    
    # Test 4: Sorting
    test_unsorted <- c(5, 2, 8, 1, 9)
    sorted_data <- sort_values(test_unsorted)
    if (sorted_data[1] == 1 && sorted_data[5] == 9) {
        cat("✓ Sorting operations: Pass\n")
    } else {
        cat("✗ Sorting operations: Fail\n")
    }
    
    # Test 5: Maximum/Minimum
    max_val <- find_maximum(test_data)
    min_val <- find_minimum(test_data)
    if (max_val == 10 && min_val == 1) {
        cat("✓ Min/Max operations: Pass\n")
    } else {
        cat("✗ Min/Max operations: Fail\n")
    }
    
    # Test 6: Group counting
    grouped_data <- c(1, 1, 2, 2, 2, 3)
    counts <- count_consecutive_groups(grouped_data)
    if (length(counts) == 3 && counts[1] == 2 && counts[2] == 3 && counts[3] == 1) {
        cat("✓ Group counting: Pass\n")
    } else {
        cat("✗ Group counting: Fail\n")
    }
    
    cat("\n")
    cat("Basic data operations validation complete\n")
}

benchmark_operations <- function(iterations = 10000) {
    # Benchmark operations for performance comparison
    cat(sprintf("Benchmarking with %d iterations...\n", iterations))
    
    # Generate larger test data
    large_data <- 1:1000
    
    # Benchmark sum calculation
    sum_time <- system.time({
        for (i in 1:iterations) {
            calculate_sum(large_data)
        }
    })["elapsed"]
    
    # Benchmark sorting
    sort_time <- system.time({
        for (i in 1:iterations) {
            sort_values(large_data)
        }
    })["elapsed"]
    
    # Benchmark filtering
    filter_time <- system.time({
        for (i in 1:iterations) {
            filter_above_threshold(large_data, 500)
        }
    })["elapsed"]
    
    cat(sprintf("Sum calculation: %.4fs\n", sum_time))
    cat(sprintf("Sorting: %.4fs\n", sort_time))
    cat(sprintf("Filtering: %.4fs\n", filter_time))
    
    return(list(
        sum_time = sum_time,
        sort_time = sort_time,
        filter_time = filter_time
    ))
}

analyze_complexity <- function() {
    # Analyze complexity of operations
    cat("Data Operations Complexity Analysis - R\n")
    cat("======================================\n")
    
    cat("Operation Complexities:\n")
    cat("  Sum/Mean calculation: O(n)\n")
    cat("  Filtering: O(n)\n")
    cat("  Sorting (introsort): O(n log n)\n")
    cat("  Min/Max finding: O(n)\n")
    cat("  Group counting: O(n)\n")
    cat("\n")
    
    cat("Memory Complexity:\n")
    cat("  Data storage: O(n)\n")
    cat("  Operation results: O(k) where k ≤ n\n")
    cat("\n")
    
    cat("R Advantages:\n")
    cat("  ✓ Built for statistical computing\n")
    cat("  ✓ Extensive statistical libraries\n")
    cat("  ✓ Vectorized operations\n")
    cat("  ✓ Rich data visualization\n")
    cat("  ✓ Interactive data exploration\n")
    cat("\n")
    
    cat("Ruchy Advantages:\n")
    cat("  ✓ Compile-time type safety\n")
    cat("  ✓ Formal mathematical verification\n")
    cat("  ✓ Systems programming capabilities\n")
    cat("  ✓ Predictable performance\n")
    cat("  ✓ Zero-cost abstractions\n")
}

main <- function() {
    # Main demonstration function
    cat("Data Operations Foundation - R\n")
    cat("=============================\n")
    cat("Statistical computing comparison with Ruchy implementation\n")
    cat("\n")
    
    # Run comprehensive test suite
    test_data_operations()
    cat("\n")
    
    # Run benchmarks
    benchmark_results <- benchmark_operations()
    cat("\n")
    
    # Analyze complexity
    analyze_complexity()
    cat("\n")
    
    cat("✅ R implementation complete\n")
    cat("📊 Ready for performance comparison with Ruchy\n")
    cat("🎯 SPRINT 23: Cross-language validation in progress\n")
}

# Run main function when script is executed
if (!interactive()) {
    main()
}